src/Microsoft.SqlTools.ServiceLayer/QueryExecution/DataStorage/SaveAsWriterBase.cs

```diff
@@ -73,7 +87,7 @@ public int WriteRow(StorageDataReader dataReader)
         /// </summary>
         /// <param name="row">The row of data to output</param>
         /// <param name="columns">The list of columns to output</param>
-        public abstract void WriteRow(IList<DbCellValue> row, IList<DbColumnWrapper> columns);
+        public abstract void WriteRow(IList<DbCellValue> row, IReadOnlyList<DbColumnWrapper> columns);
```

Comment:
This should be `IReadOnlyList<DbColumnWrapper>? columns` given that you're passed in null as a possible value on line 374 of ResultSet. 

Although I have to ask why is null something you're passing in there anyways - is there a better way you can do that such that we don't have to handle a null set of columns?

Reply:
And I assume you've made sure that the file stream writer is able to handle null columns correctly? It's not easy following the logic through a PR, what type of writer is that going to be anyways?

Reply:
> Should be `IReadOnlyList<DbColumnWrapper>?` 
Are we using nullable reference types in this project? I'm getting a warning that it has to be in a `#nullable` context if I'm going to make that change.

> Why is null something you're passing in there anyways
`SaveAsWriterBase` is only used for exporting files via SaveAs* commands. On line 374 of `ResultSet` we're generating a service buffer file, who's writer does not inherit from `SaveAsWriterBase`. I get that that's not clear from the name `fileStreamFactory` but that variable name was chosen prior to adding SaveAs support and I didn't want to go around changing anymore than necessary for this pared down PR.

The problem here is that for whatever reason (lost to time, though I vaguely remember working on this with Kevin Cunnane a long time ago) is the SaveAs writers and the service buffer writer share the same Reader/Writer factory interface despite being for significantly different purposes. The service buffer is for caching results as they come off the reader (everything is written to file, we don't know the columns at construction, writing results from a list of cell values is prohibited). The SaveAs writers are for exporting results (a selection of results can be selected, we do know columns at construction, writing results from a `DbDataReader` is prohibited). The only thing linking the two scenarios is that rows can be read from a service buffer file in both cases. The proposed refactor in the original PR was to give different reader/writer factory interfaces for SaveAs scenarios and for service buffer scenarios (and as I think on it more, I might just scrap the reader/writer factories altogether). But, that was apparently too big of a PR, so I've pared it back to just the changes necessary to get the SaveAs CSV writer to output headers on construction while not completely refactoring everything. I still intend to do that refactor as a later PR, so whatever issues come up as part of not doing those refactors will go away soon.

Reply:
Gotcha, thanks for the explanation!

(I thought we had nullable enabled here but guess not!)